File: packages/typescript/src/features/basic-capabilities/providers/time.ts
import { requireProviderSpec } from "../../../generated/spec-helpers.ts";
import type {
	IAgentRuntime,
	Memory,
	Provider,
	State,
} from "../../../types/index.ts";

// Get text content from centralized specs
const spec = requireProviderSpec("TIME");

/**
 * Time provider function that retrieves the current date and time in UTC
 * for use in time-based operations or responses.
 *
 * @param _runtime - The runtime environment of the bot agent.
 * @param _message - The memory object containing message data.
 * @returns An object containing the current date and time data, human-readable date and time string,
 * and a text response with the current date and time information.
 */
export const timeProvider: Provider = {
	name: spec.name,
	description: spec.description,
	get: async (_runtime: IAgentRuntime, _message: Memory, _state: State) => {
		const currentDate = new Date();

		// Get UTC time since bots will be communicating with users around the global
		const options = {
			timeZone: "UTC",
			dateStyle: "full" as const,
			timeStyle: "long" as const,
		};
		const humanReadable = new Intl.DateTimeFormat("en-US", options).format(
			currentDate,
		);
		return {
			data: {
				timestamp: currentDate.getTime(),
				isoString: currentDate.toISOString(),
			},
			values: {
				time: humanReadable,
			},
			text: `The current date and time is ${humanReadable}. Please use this as your reference for any time-based operations or responses.`,
		};
	},
};

File: packages/rust/src/features/basic_capabilities/bootstrap/providers/time.rs
//! TIME provider implementation.

use async_trait::async_trait;
use chrono::Utc;
use once_cell::sync::Lazy;

use crate::error::PluginResult;
use crate::generated::spec_helpers::require_provider_spec;
use crate::runtime::IAgentRuntime;
use crate::types::{Memory, ProviderResult, State};

use super::Provider;

// Get text content from centralized specs
static SPEC: Lazy<&'static crate::generated::spec_helpers::ProviderDoc> =
    Lazy::new(|| require_provider_spec("TIME"));

/// Provider for current time information (TS parity: `TIME`).
pub struct TimeProvider;

#[async_trait]
impl Provider for TimeProvider {
    fn name(&self) -> &'static str {
        &SPEC.name
    }

    fn description(&self) -> &'static str {
        &SPEC.description
    }

    fn is_dynamic(&self) -> bool {
        SPEC.dynamic.unwrap_or(true)
    }

    async fn get(
        &self,
        _runtime: &dyn IAgentRuntime,
        _message: &Memory,
        _state: Option<&State>,
    ) -> PluginResult<ProviderResult> {
        let now = Utc::now();
        let iso_string = now.to_rfc3339();
        let timestamp_ms = now.timestamp_millis();
        let human_readable = now.format("%A, %B %d, %Y at %H:%M:%S UTC").to_string();

        let text = format!(
            "The current date and time is {}. Please use this as your reference for any time-based operations or responses.",
            human_readable
        );

        Ok(ProviderResult::new(text)
            .with_value("time", human_readable)
            .with_data("timestamp", timestamp_ms)
            .with_data("isoString", iso_string))
    }
}
